### DISCUSSION TITLE
Why not also allow static virtual call-sites in generic methods constrained by abstract classes, just like we can with interfaces?

### DISCUSSION THREAD

**Discussion by daiplusplus** [dotnet/roslyn#73687] [Ideas]
(I'm asking about this after responding to [this SO question](https://stackoverflow.com/q/78526110/159145))

-----

Since C# 11, we can do this:

```
interface IHasStaticMethod
{
    public static abstract String GetFavoriteMetasyntacticVariableName();
}

class Foo : IHasStaticMethod
{
    public static String GetFavoriteMetasyntacticVariableName() => "Foo";
}

class Bar : IHasStaticMethod
{
    public static String GetFavoriteMetasyntacticVariableName() => "Bar";
}

static void Example<T>()
    where T : IHasStaticMethod
{
    String v = T.GetFavoriteMetasyntacticVariableName();
    Console.WriteLine( v );
}

Example<Foo>(); // Prints "Foo"
Example<Bar>(); // Prints "Bar"
```

I see no reason why this can't be extended to `abstract class` members either (excepting that `class` generic constraints can't be intersection types, ofc):

```
abstract class HasStaticMethod
{
    public static abstract String GetFavoriteMetasyntacticVariableName();
}

class Foo : HasStaticMethod
{
    // This causes a compiler error because Roslyn disallows the use of `static` and `override` together.
    public static override String GetFavoriteMetasyntacticVariableName() => "Foo";
}

class Bar : HasStaticMethod
{
    public static override String GetFavoriteMetasyntacticVariableName() => "Bar";
}

static void Example<T>()
    where T : HasStaticMethod
{
    String v = T.GetFavoriteMetasyntacticVariableName();
    Console.WriteLine( v );
}

Example<Foo>(); // Prints "Foo"
Example<Bar>(); // Prints "Bar"
```

...or is there a good reason for not allowing this?

**Comment by CyrusNajmabadi** [comment]
Closing out.  This is a csharplang discussion.  To answer the question:

> or is there a good reason for not allowing this?

It's possible this could be supported.  If both the runtime and language feels like it's worthwhile.  However, changes in the language need to happen starting with a discussion over at dotnet/csharplang.  dotnet/roslyn is the compiler, which implements the language as specified.

**Reply by daiplusplus** [reply]
> Closing out. This is a csharplang discussion

Ah, sorry, I forgot. I thought that was still in the Roslyn repo. My mistake.
